src/t8_cmesh/t8_cmesh_vtk_reader.cxx

```diff
@@ -65,39 +138,249 @@ t8_file_to_vtkGrid (const char *filename,
+
+    /* TODO: Avoid magic numbers in the attribute setting. */
+    /* Get and set the data of each cell */
+    for (int dtype = 0; dtype < num_data_arrays; dtype++) {
+      const t8_gloidx_t   cell_id = cell_it->GetCellId ();
+      vtkDataArray       *data = cell_data->GetArray (dtype);
+      data->GetTuple (cell_id, tuples[dtype]);
+      t8_cmesh_set_attribute (cmesh, tree_id, t8_get_package_id (), dtype + 1,
+                              tuples[dtype], data_size[dtype], 0);
+    }
+    /* Check geometry-dimension */
+    if (max_dim < cell_it->GetCellDimension ()) {
+      max_dim = cell_it->GetCellDimension ();
+    }
+    tree_id++;
+  }
+  t8_debugf ("[D] read %li trees\n", tree_id);
+
+  /* Clean-up */
+  cell_it->Delete ();
+  if (num_data_arrays > 0) {
+    T8_FREE (data_size);
+    for (int idata = num_data_arrays - 1; idata >= 0; idata--) {
+      T8_FREE (tuples[idata]);
+    }
+    T8_FREE (tuples);
+  }
+  T8_FREE (vertices);
+  return tree_id;
+}
+
+/**
+ * Given a pointer to a vtkDataSet a cmesh representing the vtkDataSet is
+ * constructed and can be shared over the processes. 
+ * 
+ * \param[in] vtkGrid A pointer to a vtkDataSet
+ * \param[in] partition Flag if the cmesh should be partitioned
+ * \param[in] main_proc The main reading process
+ * \param[in] comm The communicator. 
+ * \return t8_cmesh_t 
+ */
+t8_cmesh_t
+t8_vtkGrid_to_cmesh (vtkSmartPointer < vtkDataSet > vtkGrid,
+                     const int partition, const int main_proc,
+                     sc_MPI_Comm comm)
+{
+  t8_cmesh_t          cmesh;
+  int                 mpisize;
+  int                 mpirank;
+  int                 mpiret;
+  /* Get the size of the communicator and the rank of the process. */
+  mpiret = sc_MPI_Comm_size (comm, &mpisize);
+  SC_CHECK_MPI (mpiret);
+  mpiret = sc_MPI_Comm_rank (comm, &mpirank);
+  SC_CHECK_MPI (mpiret);
+
+  /* Ensure that the main-proc is a valid proc. */
+  T8_ASSERT (0 <= main_proc && main_proc < mpisize);
+
+  /* Already declared here, because we might use them during communication */
+  t8_gloidx_t         num_trees;
+  int                 dim;
+
+  t8_cmesh_init (&cmesh);
+  if (!partition || mpirank == main_proc) {
+    num_trees = t8_vtk_iterate_cells (vtkGrid, cmesh, comm);
+    dim = t8_get_dimension (vtkGrid);
+    t8_cmesh_set_dimension (cmesh, dim);
+    t8_geometry_c      *linear_geom = t8_geometry_linear_new (dim);
+    t8_cmesh_register_geometry (cmesh, linear_geom);
+  }
+  if (partition) {
+    t8_gloidx_t         first_tree;
+    t8_gloidx_t         last_tree;
+    if (mpirank == main_proc) {
+      first_tree = 0;
+      last_tree = num_trees - 1;
+    }
+    /* Communicate the dimension to all processes */
+    sc_MPI_Bcast (&dim, 1, sc_MPI_INT, main_proc, comm);
+    t8_debugf ("[D] dim: %i\n", dim);
+    /* Communicate the number of trees to all processes. 
+     * TODO: This probably crashes when a vtkGrid is distributed in many 
```

Comment:
Oh, oh ...

Reply:
This is left intentionally, because we will extend the reader to read parallel vtk-Data in a (near-) future-branch. 
I updated the comment to explain the idea that will solve it.